pkg/client/cache/expiration_cache.go

```diff
@@ -157,6 +162,9 @@ func (c *ExpirationCache) Update(obj interface{}) error {
 
 // Delete removes an item from the cache.
 func (c *ExpirationCache) Delete(obj interface{}) error {
+	// We don't need to take the expiration lock here because the caller wants
+	// to delete the given key, and the worst that can happen is that it already
+	// expired.
```

Comment:
the cacheStorage.Delete is still locked (internally in the thread safe store). Delete just removes it from a dict/store, it won't get GC'd till everything with a reference lets go. Calling delete on a key that isn't in the store should be ok.